File: apps/intellij/src/main/kotlin/dev/nx/console/graph/NxGraphServer.kt
import com.intellij.openapi.Disposable
import com.intellij.openapi.components.Service
import com.intellij.openapi.diagnostic.logger
import com.intellij.openapi.diagnostic.thisLogger
import com.intellij.openapi.project.Project
import com.intellij.util.io.readLineAsync
import dev.nx.console.graph.NxGraphRequest
import dev.nx.console.nxls.NxWorkspaceRefreshListener
import dev.nx.console.nxls.NxlsService
import dev.nx.console.utils.NxGeneralCommandLine
import java.io.IOException
import java.net.InetAddress
import java.net.ServerSocket
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import kotlinx.coroutines.*

data class WebviewRequest(val type: String, val id: String) {}

data class WebviewResponse(val type: String, val id: String, val payload: String) {}

@Service(Service.Level.PROJECT)
class StandardNxGraphServer(project: Project, cs: CoroutineScope) :
    NxGraphServer(project, 5580, false, cs) {
    companion object {
        fun getInstance(project: Project): NxGraphServer =
            project.getService(StandardNxGraphServer::class.java)
    }
}

@Service(Service.Level.PROJECT)
class AffectedNxGraphServer(project: Project, cs: CoroutineScope) :
    NxGraphServer(project, 5590, true, cs) {
    companion object {
        fun getInstance(project: Project): NxGraphServer =
            project.getService(AffectedNxGraphServer::class.java)
    }
}

val logger = logger<NxGraphServer>()

open class NxGraphServer(
    private val project: Project,
    private val startPort: Int,
    private val affected: Boolean,
    private val cs: CoroutineScope
) : Disposable {

    var currentPort: Int? = null
    private var nxGraphProcess: Process? = null

    private var lastErrror: String? = null

    private var isStarted = false
    private var isStarting = false

    init {
        with(project.messageBus.connect(this)) {
            subscribe(
                NxlsService.NX_WORKSPACE_REFRESH_TOPIC,
                NxWorkspaceRefreshListener {
                    nxGraphProcess?.apply {
                        if (!isAlive) {
                            start()
                        }
                    }
                }
            )
        }
    }

    suspend fun handleGraphRequest(request: NxGraphRequest, attempt: Int = 0): NxGraphRequest {
        try {

            if (nxGraphProcess == null || nxGraphProcess?.isAlive != true && !isStarting) {
                start()
                waitForServerReady()
            }
            if (!isStarted) {
                waitForServerReady()
            }

            var url = "http://localhost:${this.currentPort}/"
            url +=
                when (request.type) {
                    "requestProjectGraph" -> "project-graph.json"
                    "requestTaskGraph" -> "task-graph.json"
                    "requestExpandedTaskInputs" -> "task-inputs.json?taskId=${request.payload}"
                    "requestSourceMaps" -> "source-maps.json"
                    else -> throw Exception("unknown request type ${request.type}")
                }

            val response =
                withContext(Dispatchers.IO) {
                    val client = HttpClient.newBuilder().build()

                    val httpRequest =
                        HttpRequest.newBuilder()
                            .uri(URI.create(url))
                            .header("Accept-Encoding", "gzip, deflate")
                            .build()
                    client.send(httpRequest, HttpResponse.BodyHandlers.ofString())
                }
            return NxGraphRequest(type = request.type, id = request.id, payload = response.body())
        } catch (e: Throwable) {
            logger.info("error while handling graph request: $e")
            if (attempt == 0) {
                return handleGraphRequest(request, 1)
            }

            val error =
                if (e is TimeoutCancellationException && lastErrror != null)
                    "error while running nx graph: $lastErrror"
                else e.message
            return NxGraphRequest(
                type = request.type,
                id = request.id,
                payload = request.payload,
                error = error
            )
        }
    }

    fun start() {
        if ((isStarted && nxGraphProcess?.isAlive == true) || isStarting) {
            return
        }

        isStarting = true
        isStarted = false
        cs.launch {
            var port = startPort
            var isPortAvailable = false
            while (!isPortAvailable) {
                isPortAvailable = checkPort(port)
                if (!isPortAvailable) {
                    port++
                }
            }

            currentPort = port
            try {
                val process = spawnProcess(port)
                nxGraphProcess = process
                handleGraphProcessError(process)

                isStarted = true
                isStarting = false
            } catch (e: Exception) {
                println("error while starting nx graph: $e")
                isStarting = false
            }
        }
    }

    private suspend fun spawnProcess(port: Int): Process {
        println("trying to start graph at $port")

        return withContext(Dispatchers.IO) {
            val commandLine =
                NxGeneralCommandLine(
                    project,
                    listOf(
                        "graph",
                        "--port",
                        port.toString(),
                        "--open",
                        "false",
                        "--watch",
                        if (affected) "--affected" else ""
                    ),
                )

            val process = commandLine.createProcess()
            if (!process.isAlive) {
                throw IOException("Unable to start graph at port $port")
            } else {
                logger.info("graph server started: $process")
            }

            // wait for process to start - signified by logging the port
            val reader = process.inputStream.bufferedReader()
            var stopWaiting = false
            var line: String? = null
            line = reader.readLineAsync()
            thisLogger().trace("Read line while starting: $line")
            while (!stopWaiting) {
                if (line != null && line.contains(port.toString())) {
                    stopWaiting = true
                }
                line = reader.readLineAsync()?.trim()?.lowercase()
                thisLogger().trace("Read line while starting: $line")
            }

            process
        }
    }

    private fun checkPort(port: Int): Boolean {
        return try {
            ServerSocket(port, 1, InetAddress.getByName("127.0.0.1")).use {
                return true
            }
        } catch (e: IOException) {
            false
        }
    }

    private fun handleGraphProcessError(process: Process) {
        lastErrror = null
        process.onExit().thenAccept {
            logger.debug("graph server exited with code ${it.exitValue()}")
            isStarted = false
            isStarting = false
            nxGraphProcess = null

            if (it.exitValue() != 0) {
                val stdErr = process.errorStream.readAllBytes().decodeToString()
                val stdOut = process.inputStream.readAllBytes().decodeToString()
                logger.debug(
                    "graph server exited with error. \n stdErr: $stdErr \n stdOut: $stdOut"
                )
                lastErrror =
                    if (stdErr != null && stdErr.length > 0) stdErr
                    else "graph exited unexpectedly. Full logs: $stdOut"
            }
        }
    }

    private suspend fun waitForServerReady() {
        withTimeout(10000) {
            while (!isStarted) {
                delay(100)
            }
        }
    }

    override fun dispose() {
        nxGraphProcess?.destroyForcibly()
        nxGraphProcess = null
        isStarted = false
        isStarting = false
    }
}

interface NxGraphServerRefreshListener {
    fun onRefresh()
}

File: libs/vscode/graph-base/src/lib/nx-graph-server.ts
import { getNxWorkspacePath } from '@nx-console/vscode/nx-workspace';
import { ChildProcess, spawn } from 'child_process';
import { createServer } from 'net';
import { getPackageManagerCommand } from 'nx/src/devkit-exports';
import { xhr } from 'request-light';
import { Disposable, EventEmitter, ExtensionContext } from 'vscode';

let nxGraphServer: NxGraphServer | undefined = undefined;
let nxGraphServerAffected: NxGraphServer | undefined = undefined;

export function getNxGraphServer(context: ExtensionContext, affected = false) {
  if (affected) {
    if (!nxGraphServerAffected) {
      nxGraphServerAffected = new NxGraphServer(5567, affected);
      nxGraphServerAffected.start();
      context.subscriptions.push(nxGraphServerAffected);
    }
    return nxGraphServerAffected;
  }
  if (!nxGraphServer) {
    nxGraphServer = new NxGraphServer(5566);
    nxGraphServer.start();
    context.subscriptions.push(nxGraphServer);
  }

  return nxGraphServer;
}

export class NxGraphServer implements Disposable {
  private currentPort: number | undefined = undefined;
  private nxGraphProcess: ChildProcess | undefined = undefined;

  isStarting = false;
  isStarted = false;
  updatedEventEmitter = new EventEmitter();

  constructor(private startPort: number, private affected = false) {}

  async handleWebviewRequest(request: {
    type: string;
    id: string;
    payload: string;
  }): Promise<
    | {
        type: string;
        id: string;
        payload: string;
      }
    | undefined
  > {
    try {
      if (this.isCrashed) {
        await this.start();
      }
      if (!this.isStarted) {
        await this.waitForServerReady();
      }

      const { type, id } = request;

      let url = `http://localhost:${this.currentPort}/`;
      switch (type) {
        case 'requestProjectGraph':
          url += 'project-graph.json';
          break;
        case 'requestTaskGraph':
          url += 'task-graph.json';
          break;
        case 'requestExpandedTaskInputs':
          url += `task-inputs.json?taskId=${request.payload}`;
          break;
        case 'requestSourceMaps':
          url += 'source-maps.json';
          break;
        default:
          return;
      }

      const headers = { 'Accept-Encoding': 'gzip, deflate' };
      const response = await xhr({
        url,
        headers,
      });
      const data = response.responseText;
      return {
        type: `${type}Response`,
        id,
        payload: data,
      };
    } catch (error) {
      console.log('error while handling webview request', error);
      return;
    }
  }

  /**
   * starts nx graph server
   */
  async start(): Promise<{ error: string } | undefined> {
    if (this.isStarting) {
      return;
    }
    if (this.isStarted && !this.isCrashed) {
      return;
    }
    this.isStarted = false;
    this.isStarting = true;
    let port = this.startPort;

    let isPortAvailable = false;
    while (!isPortAvailable) {
      isPortAvailable = await this.checkPort(port);
      if (!isPortAvailable) {
        port++;
      }
    }

    this.currentPort = port;
    try {
      await this.spawnProcess(port);
      console.log('successfully started nx graph at port', port);
      this.isStarted = true;
      this.isStarting = false;
    } catch (error) {
      console.error(`error while starting nx graph: ${error}`);
      this.isStarting = false;
      this.isStarted = false;
      return { error: `${error}` };
    }
  }

  private get isCrashed() {
    return !!this.nxGraphProcess && !!this.nxGraphProcess.exitCode;
  }

  private async spawnProcess(port: number): Promise<void> {
    const workspacePath = await getNxWorkspacePath();

    return new Promise((resolve, reject) => {
      const nxGraphProcess = spawn(
        getPackageManagerCommand().exec,
        [
          'nx',
          'graph',
          `--port`,
          `${port}`,
          '--open',
          'false',
          '--watch',
          this.affected ? '--affected' : '',
        ],
        {
          cwd: workspacePath,
          windowsHide: true,
          shell: true,
          env: process.env,
        }
      );

      nxGraphProcess.stdout.setEncoding('utf8');
      nxGraphProcess.stderr.setEncoding('utf8');

      let stdErrOutput = '';
      nxGraphProcess.stdout.on('data', (data) => {
        const text: string = data.toString().trim().toLowerCase();

        if (!text) return;
        if (text.includes(`${port}`)) {
          resolve();
          return;
        }
        if (text.includes('updated')) {
          this.updatedEventEmitter.fire(true);
        }
      });
      nxGraphProcess.stderr.on('data', (data) => {
        stdErrOutput += data.toString();
      });
      nxGraphProcess.on('exit', async () => {
        this.isStarted = false;
        reject(stdErrOutput);
      });

      this.nxGraphProcess = nxGraphProcess;
    });
  }

  private checkPort(port: number): Promise<boolean> {
    return new Promise((resolve) => {
      const server = createServer();

      server.listen(port, '127.0.0.1');
      server.on('listening', () => {
        server.close();
        resolve(true);
      });

      server.on('error', () => {
        resolve(false);
      });
    });
  }

  private waitForServerReady(): Promise<void> {
    return new Promise((resolve, reject) => {
      const timeout = 10000;
      const checkInterval = setInterval(async () => {
        if (this.isStarted) {
          clearInterval(checkInterval);
          clearTimeout(timeoutId);
          resolve();
        }
        if (this.isCrashed) {
          clearTimeout(timeoutId);
          clearInterval(checkInterval);
          reject(new Error('Server crashed during startup'));
        }
      }, 100);

      const timeoutId = setTimeout(() => {
        clearInterval(checkInterval);
        reject(new Error('Server did not start within 10 seconds'));
      }, timeout);
    });
  }

  dispose() {
    this.nxGraphProcess?.kill();
    this.nxGraphProcess = undefined;
  }
}
